fedimint-server/src/config/io.rs

```diff
@@ -113,11 +119,180 @@ fn plaintext_display_write<T: Display>(obj: &T, path: &Path) -> anyhow::Result<(
 }
 
 /// Writes struct into an encrypted json file
-fn encrypted_json_write<T: Serialize + DeserializeOwned>(
+pub fn encrypted_json_write<T: Serialize + DeserializeOwned>(
     obj: &T,
     key: &LessSafeKey,
     path: &Path,
 ) -> anyhow::Result<()> {
     let bytes = serde_json::to_string(obj)?.into_bytes();
     encrypted_write(bytes, key, path.with_extension(ENCRYPTED_EXT))
 }
+
+/// We definitely don't want leading/trailing newlines in passwords, and a user
+/// editing the file manually will probably get a free newline added
+/// by the text editor.
+pub fn trim_password(password_untrimmed: &str) -> &str {
+    let password = password_untrimmed.trim_matches('\n');
+    // In the future we also don't want to support any leading/trailing newlines
+    let password_fully_trimmed = password.trim();
+    if password_fully_trimmed != password {
+        warn!(
+            target: LOG_CORE,
+            "Password in the password file contains leading/trailing whitespaces. This will an error in the future."
+        );
+    }
+    password_fully_trimmed
+}
+
+pub fn backup_copy_path(original: &Path) -> PathBuf {
+    original.with_extension("bak")
+}
+
+pub fn create_backup_copy(original: &Path) -> anyhow::Result<()> {
+    let backup_path = backup_copy_path(original);
+    info!(target: LOG_CORE, ?original, ?backup_path, "Creating backup copy of file");
+    ensure!(
+        !backup_path.exists(),
+        "Already have a backup at {backup_path:?}, would be overwritten"
+    );
+    fs::copy(original, backup_path)?;
+    Ok(())
+}
+
+/// Re-encrypts the private config with a new password.
+///
+/// Note that we assume that the in-memory secret config equals the on-disk
+/// secret config. If the process is interrupted,
+/// [`recover_interrupted_password_change`] will fix it on startup.
+///
+/// As an additional safetynet this function creates backup copies of all files
+/// being overwritten. These will be deleted by [`finalize_password_change`]
+/// after the config has been read successfully for the first time after a
+/// password change.
+pub fn reencrypt_private_config(
+    data_dir: &Path,
+    private_config: &ServerConfigPrivate,
+    new_password: &str,
+) -> anyhow::Result<()> {
+    info!(target: LOG_CORE, ?data_dir, "Re-encrypting private config with new password");
+    let trimmed_password = trim_password(new_password);
+
+    // we keep the same salt so we don't have to atomically update 3 files, 2 is
+    // annoying enough (if we have to write the password file)
+    let salt = fs::read_to_string(data_dir.join(SALT_FILE))?;
```

Comment:
Apologies for the early review while this is still in draft.

The extra hassle may be worth it since an attacker can reuse any work done to generate a rainbow table with the old salt on the new password.

Reply:
From my perspective the difference in security between using no/a very commonly used salt and static salt per federation is way higher than the difference between static and re-generated salt. Generating rainbow tables to crack a single password seems a bit more exotic than using pre-generated ones for common crypto schemes. 

I'm open to change it if consensus is it's worth it, but it does seem like a very incremental security improvement to me (even though the complexity hit is likely lower than I thought when writing the comment).

Reply:
Those are all valid points :+1: 

After thinking through some of the additional changes required for handling a new salt, I agree the additional complexity isn't worth it.